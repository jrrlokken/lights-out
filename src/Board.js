import React, { useState } from 'react';
import Cell from './Cell'
import './Board.css'

function Board({ nrows = 5, ncols = 5, chanceLightStartsOn = 0.25 }) {
  const [board, setBoard] = useState(createBoard());

  function createBoard() {
    let initialBoard = [];

    for (let y=0; y < nrows; y++) {
      let row = [];
      for (let x = 0; x < ncols; x++) {
        row.push(Math.random() < chanceLightStartsOn)
      }
      initialBoard.push(row);
    }
    return initialBoard;
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split('-').map(Number);

      const flipCell = (y, x, boardCopy) => {
        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      const boardCopy = oldBoard.map(row => [...row]);

      flipCell(y, x, boardCopy);
      flipCell(y, x - 1, boardCopy);
      flipCell(y, x + 1, boardCopy);
      flipCell(y - 1, x, boardCopy);
      flipCell(y + 1, x, boardCopy);

      return boardCopy;

    })
  }

  function hasWon() {
    return board.every(row => row.every(cell => !cell));
  }

  if (hasWon()) {
    return (
      <div className="Board-title">
        <div className="Winner">
          <span className="neon-orange">YOU</span>
          <span className="neon-blue">WIN!</span>
        </div>
      </div>
    ) 
  }

  let tblBoard = [];

  for(let y = 0; y < nrows; y++) {
    let row = [];
    for(let x = 0; x < ncols; x++) {
      let coord = `${y}-${x}`;
      row.push(
        <Cell 
          key={coord} 
          isLit={board[y][x]}
          flipCellsAroundMe={() => flipCellsAround(coord)}
        />
      );
    }
    tblBoard.push(<tr key={y}>{row}</tr>);
  }


  return (
    <div>
      <div className="Board-title">
        <div className="neon-orange">Lights</div>
        <div className="neon-blue">Out</div>
      </div>
        <table className="Board">
          <tbody>{tblBoard}</tbody>
        </table>
    </div>
  )
}

export default Board;