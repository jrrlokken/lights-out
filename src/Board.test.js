import React from "react";
import { render } from "@testing-library/react";
import Board from "./Board";

// smoke test
it("renders without crashing", function() {
  render(<Board />);
});

// snapshot test
it("matches snapshot", function() {
  const {asFragment} = render(<Board />);
  expect(asFragment()).toMatchSnapshot();
});