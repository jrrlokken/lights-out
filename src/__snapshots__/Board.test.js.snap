// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="Board-title"
    >
      <div
        class="neon-orange"
      >
        Lights
      </div>
      <div
        class="neon-blue"
      >
        Out
      </div>
    </div>
    <table
      class="Board"
    >
      <tbody>
        <tr>
          <td
            class="Cell"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
        </tr>
        <tr>
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell Cell-lit"
          />
        </tr>
        <tr>
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
          <td
            class="Cell"
          />
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
